<a href="https://colab.research.google.com/github/Hendro10/TUGAS2_MACHINE_LEARNING_NIM-200401072103_NAMA_HENDRO_GUNAWAN_KELAS_IT-602/blob/main/Aturan_Asosiasi_(Algoritma_Apriori).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

APRIORI

1. Import Library

In [ ]:
!pip install apyori

  Preparing metadata (setup.py) ... done
  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5956 sha256=7a65336ac446038b4191b4ef2652762405ebc79be6baf62321c8db3d5b0f7533
  Stored in directory: /root/.cache/pip/wheels/c4/1a/79/20f55c470a50bb3702a8cb7c94d8ada15573538c7f4baebe2d
Successfully built apyori


In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

2. Pra Proses Data

In [ ]:
# Mounted At/Content/Drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
dataset = pd.read_csv('/content/drive/MyDrive/KuliahUNSIA/Semester5/MachineLearning/Pertemuan11/Grocery_Store_DataSet.csv', header = None)
transactions = []
for i in range(0, 7501):
  transactions.append([str(dataset.values[i,j]) for j in range(0, 20)])

In [ ]:
print(dataset)

           0                    1                    2                 3   \
0     Item(s)               Item 1               Item 2            Item 3   
1           4         citrus fruit  semi-finished bread         margarine   
2           3       tropical fruit               yogurt            coffee   
3           1           whole milk                  NaN               NaN   
4           4            pip fruit               yogurt      cream cheese   
...       ...                  ...                  ...               ...   
9831       17              sausage              chicken              beef   
9832        1    cooking chocolate                  NaN               NaN   
9833       10              chicken         citrus fruit  other vegetables   
9834        4  semi-finished bread        bottled water              soda   
9835        5              chicken       tropical fruit  other vegetables   

                  4              5               6                7   \
0  

3. Menampilkan Sebanyak 5 Data Teratas

In [ ]:
df= dataset = pd.read_csv('/content/drive/MyDrive/KuliahUNSIA/Semester5/MachineLearning/Pertemuan11/Grocery_Store_DataSet.csv', header = None)
df.head(5)

0               1                    2             3             4   \
0  Item(s)          Item 1               Item 2        Item 3        Item 4   
1        4    citrus fruit  semi-finished bread     margarine   ready soups   
2        3  tropical fruit               yogurt        coffee           NaN   
3        1      whole milk                  NaN           NaN           NaN   
4        4       pip fruit               yogurt  cream cheese  meat spreads   

       5       6       7       8       9   ...       23       24       25  \
0  Item 5  Item 6  Item 7  Item 8  Item 9  ...  Item 23  Item 24  Item 25   
1     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   
2     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   
3     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   
4     NaN     NaN     NaN     NaN     NaN  ...      NaN      NaN      NaN   

        26       27       28       29       30       31       32  
0  Item 26  Item 27  Item 28  Item 29  Item 30  Item 31  Item 32  
1      NaN      NaN      NaN      NaN      NaN      NaN      NaN  
2      NaN      NaN      NaN      NaN      NaN      NaN      NaN  
3      NaN      NaN      NaN      NaN      NaN      NaN      NaN  
4      NaN      NaN      NaN      NaN      NaN      NaN      NaN  

[5 rows x 33 columns]

4. Melihat Variabel dalam Dataset

In [ ]:
df.keys()

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32],
           dtype='int64')

5. Melihat Shape

In [ ]:
df.shape

(9836, 33)

Di sini terlihat bahwa terdapat 9836 baris dan 33 kolom.

6. Training Apriori

In [ ]:
from apyori import apriori
rules = apriori(transactions = transactions, min_support = 0.003, min_confidence = 0.2, min_lift = 3, min_length = 2, max_length = 2)

7. Menampilkan Hasil Training

In [ ]:
results = list(rules)
results: list

[RelationRecord(items=frozenset({'10', 'fruit/vegetable juice'}), support=0.0061325156645780565, ordered_statistics=[OrderedStatistic(items_base=frozenset({'10'}), items_add=frozenset({'fruit/vegetable juice'}), confidence=0.25136612021857924, lift=3.46598762455802)]),
 RelationRecord(items=frozenset({'10', 'margarine'}), support=0.005465937874950006, ordered_statistics=[OrderedStatistic(items_base=frozenset({'10'}), items_add=frozenset({'margarine'}), confidence=0.2240437158469945, lift=3.734559805707346)]),
 RelationRecord(items=frozenset({'10', 'root vegetables'}), support=0.008132249033462205, ordered_statistics=[OrderedStatistic(items_base=frozenset({'10'}), items_add=frozenset({'root vegetables'}), confidence=0.3333333333333333, lift=3.0160836349014875)]),
 RelationRecord(items=frozenset({'11', 'domestic eggs'}), support=0.0038661511798426876, ordered_statistics=[OrderedStatistic(items_base=frozenset({'11'}), items_add=frozenset({'domestic eggs'}), confidence=0.21804511278195488,

8. Merubah Dataset Kedalam Bentuk Kolom

In [ ]:
def inspect(results):
    lhs         = [tuple(result[2][0][0])[0] for result in results]
    rhs         = [tuple(result[2][0][1])[0] for result in results]
    supports    = [result[1] for result in results]
    confidences = [result[2][0][2] for result in results]
    lifts       = [result[2][0][3] for result in results]
    return list(zip(lhs, rhs, supports, confidences, lifts))
resultsinDataFrame = pd.DataFrame(inspect(results), columns = ['Left Hand Side', 'Right Hand Side', 'Support', 'Confidence', 'Lift'])

9. Melihat Hasil

In [ ]:
resultsinDataFrame: pd.DataFrame

Left Hand Side        Right Hand Side   Support  Confidence      Lift
0                 10  fruit/vegetable juice  0.006133    0.251366  3.465988
1                 10              margarine  0.005466    0.224044  3.734560
2                 10        root vegetables  0.008132    0.333333  3.016084
3                 11          domestic eggs  0.003866    0.218045  3.379249
4                 11              margarine  0.003866    0.218045  3.634570
5                 12          domestic eggs  0.003466    0.276596  4.286663
6                 12  fruit/vegetable juice  0.003200    0.255319  3.520494
7                 12       other vegetables  0.007599    0.606383  3.156474
8                 12              pip fruit  0.003333    0.265957  3.445504
9                 12        root vegetables  0.004399    0.351064  3.176514
10                12         tropical fruit  0.004133    0.329787  3.139256
11                12     whipped/sour cream  0.003333    0.265957  3.640414
12                13     whipped/sour cream  0.003066    0.377049  5.161033
13                13                 yogurt  0.003733    0.459016  3.342798
14                14       other vegetables  0.004933    0.649123  3.378952
15                14        root vegetables  0.003466    0.456140  4.127272
16                14                 yogurt  0.003333    0.438596  3.194090
17     baking powder                  sugar  0.003733    0.208955  6.269493
18     baking powder     whipped/sour cream  0.005066    0.283582  3.881659
19              beef        root vegetables  0.017598    0.346457  3.134827
20           berries     whipped/sour cream  0.007732    0.250000  3.421989
21            liquor           bottled beer  0.004533    0.409639  5.087250
22    red/blush wine           bottled beer  0.004933    0.262411  3.258854
23             flour              margarine  0.004133    0.234848  3.914663
24             flour                  sugar  0.005066    0.287879  8.637515
25             flour     whipped/sour cream  0.004133    0.234848  3.214596
26           mustard            frankfurter  0.003066    0.230000  3.920977
27  processed cheese  fruit/vegetable juice  0.003466    0.218487  3.012636
28            grapes         tropical fruit  0.005999    0.326087  3.104033
29  processed cheese                    ham  0.003333    0.210084  8.207502
30               ham            white bread  0.005333    0.208333  4.823174
31             herbs        root vegetables  0.006532    0.404959  3.664168
32  processed cheese            white bread  0.004799    0.302521  7.003735
33              rice        root vegetables  0.003200    0.406780  3.680644
34     sliced cheese                sausage  0.006932    0.300578  3.239419
35            turkey         tropical fruit  0.003066    0.348485  3.317240

10. Kesimpulan

Berdasarkan hasil perhitungan menggunakan python di atas terdapat 19 (sembilan belas) macam pilihan, yaitu diantaranya: pembeli yang membeli (bubuk pengembang)kemungkinan akan membeli sugar (gula), pembeli yang membeli baking powder (bubuk pengembang) kemungkinan juga akan membeli  (krim kue), pembeli yang membeli  (daging sapi) kemungkinan akan membeli root vegetables (sayuran), pembeli yang membeli berries (buah beri) kemungkinan akan membeli whipped/sour cream (krim kue), pembeli yang membeli red/blushwine (anggur merah) kemungkinan akan membeli bottled beer (bir botol), pembeli yang membeli flour (tepung) kemungkinan akan membeli margarine (mentega), pembeli yang membeli flour (tepung) kemungkinan akan membeli sugar (gula), pembeli yang membeli flour (tepung) kemungkinan akan membeli whipped/sour cream (krim kue), pembeli yang membeli mustard (saus moster) kemungkinan akan membeli frankfurter (sosis), pembeli yang membeli processed cheese (keju) kemungkinan akan membeli fruit/vegetable juice (jus sayur/buah), pembeli yang membeli grape (anggur) kemungkinan akan membeli tropical fruit (buah tropis), pembeli yang membeli processed cheese (keju) kemungkinan akan membeli ham (daging babi yang dikeringkan), pembeli yang membeli ham (daging babi yang dikeringkan) kemungkinan akan membeli white bread (roti putih), pembeli yang membeli rice (beras) kemungkinan akan membeli root vegetables (sayuran), pembeli yang membeli sliced cheese (irisan keju) kemungkinan akan membeli sausage (sosis), dan pembeli yang membeli turkey (ayam kalkun) kemungkinan akan membeli tropical fruit (buah tropis).

Terima kasih